trees/src/iavl/tree.rs

```diff
@@ -124,19 +171,174 @@ impl Default for Node {
+            };
+
+            let t2 = y.left_node;
+
+            // Perform rotation on z and update height and hash
+            z.right_node = t2;
+            z.right_hash = y.left_hash;
+            z.update_height_and_size_get_balance_factor(node_db);
+            z.version = version;
+            let z = Node::Inner(z);
+
+            // Perform rotation on y, update hash and update height
+            y.left_hash = z.hash();
+            y.left_node = Some(Box::new(z));
+            y.update_height_and_size_get_balance_factor(node_db);
+            y.version = version;
+
+            *self = Node::Inner(y);
+
+            Ok(())
+        } else {
+            // Can't rotate a leaf node
+            Err(Error::RotateError(LEAF_ROTATE_ERROR.to_owned()))
+        }
+    }
+
+    /// Updates the node's height and size and balance factor
+    /// Balances a node if it's balance factor is between -2 and 2 (inclusive)
+    fn update_height_and_size_and_balance<T: Database>(
+        &mut self,
+        version: u32,
+        node_db: &NodeDB<T>,
+    ) -> Result<(), Error> {
+        match self {
+            Node::Leaf(_) => {
+                // A leaf node is always balanced
+                Ok(())
+            }
+            Node::Inner(inner) => match inner.update_height_and_size_get_balance_factor(node_db) {
+                -2 => {
+                    let right_node = inner.get_mut_right_node(node_db);
+
+                    if right_node.get_balance_factor(node_db) <= 0 {
+                        return Ok(Self::left_rotate(self, version, node_db)
+                            .expect("given the imbalance, expect rotation to always succeed"));
+                    }
+
+                    Self::right_rotate(right_node, version, node_db)
+                        .expect("given the imbalance, expect rotation to always succeed");
+                    Ok(Self::left_rotate(self, version, node_db)
+                        .expect("given the imbalance, expect rotation to always succeed"))
+                }
+
+                2 => {
+                    let left_node = inner.get_mut_left_node(node_db);
+
+                    if left_node.get_balance_factor(node_db) >= 0 {
+                        return Ok(Self::right_rotate(self, version, node_db)
+                            .expect("given the imbalance, expect rotation to always succeed"));
+                    }
+
+                    Self::left_rotate(left_node, version, node_db)
+                        .expect("given the imbalance, expect rotation to always succeed");
+                    Ok(Self::right_rotate(self, version, node_db)
+                        .expect("given the imbalance, expect rotation to always succeed"))
+                }
+                -1..=1 => {
+                    // The node is balanced
+                    Ok(())
+                }
+                _ => Err(Error::Balancing),
```

Comment:
You could panic here, 'cause you already do this + this state should be unreachable.

Reply:
Or instead of `.expect`, you could return error too

Reply:
It's only at the point the method is called can we be certain that we won't reach here. From within the method we cannot tell. This is different to the other panics in this method, where we know the rotation will succeed because we've checked (within this method) the tree's balance factor.

Reply:
Tree must be rotated when balance == -2/2. Otherwise, this is an issue

Reply:
For now, it's okay to leave this as is, but I think we could bring this topic later. Maybe even change interface of remove method for returning result without any panic inside